File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 092335. k진수에서 소수 개수 구하기/kotlin/Solution.kt
import kotlin.math.sqrt

class Solution {
    fun solution(n: Int, k: Int): Int {
        val s = n.toString(k)
        var answer = 0
        var v = 0L
        for (i in 0.. s.length) {
            if (i == s.length || s[i] == '0') {
                if (v > 0) {
                    if (isPrime(v)) answer ++
                    v = 0
                }
            } else {
                v = v * 10 + s[i].code - 48
            }
        }
        return answer
    }
}

private fun isPrime(num: Long):Boolean {
    if (num == 1L) {
        return false
    }
    for (i in 2L..sqrt(num.toDouble()).toLong()) {
        if (num%i == 0L) {
            return false
        }
    }
    return true
}

File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 092335. k진수에서 소수 개수 구하기/python/solution.py
from math import sqrt

def solution(n, k):
    return len([num for num in split_numbers(n, k) if is_prime_number(num)])


# n진법 변환 후, 0이 아닌 숫자들 분리
def split_numbers(n, k):
    stack = [] # First In Last Out
    while n > 0:
        n, r = divmod(n, k)
        stack.append(r)
    cur = 0
    result = []
    while stack:
        head = stack.pop()
        if head != 0:
            cur *= 10
            cur += head
        elif cur != 0:
            result.append(cur)
            cur = 0
    if cur != 0:
        result.append(cur)
    return result


# 소수 여부 확인
def is_prime_number(n):
    if n == 1:
        return False
    for i in range(2, int(sqrt(n)) + 1):
        if n % i == 0:
            return False
    return True
